tokio/src/sync/broadcast.rs

```diff
@@ -804,6 +804,47 @@ impl<T> Sender<T> {
         Arc::ptr_eq(&self.shared, &other.shared)
     }
 
+    /// A future which completes when the number of [Receiver]s subscribed to this `Sender` reaches
+    /// zero.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use futures::FutureExt;
+    /// use tokio::sync::broadcast;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx1) = broadcast::channel::<u32>(16);
+    ///     let mut rx2 = tx.subscribe();
+    ///
+    ///     tokio::spawn(async move {
+    ///         assert_eq!(rx1.recv().await.unwrap(), 10);
+    ///     });
+    ///
+    ///     let _ = tx.send(10);
+    ///     assert!(tx.closed().now_or_never().is_none());
+    ///
+    ///     let _ = tokio::spawn(async move {
+    ///         assert_eq!(rx2.recv().await.unwrap(), 10);
+    ///     }).await;
+    ///
+    ///     assert!(tx.closed().now_or_never().is_some());
+    /// }
+    /// ```
+    pub async fn closed(&self) {
+        std::future::poll_fn(|_| {
+            let tail = self.shared.tail.lock();
+
+            if tail.closed || tail.rx_cnt == 0 {
+                Poll::Ready(())
+            } else {
+                Poll::Pending
+            }
+        })
+        .await;
+    }
```

Comment:
I think this future will only get polled once because you are not waking up the task again.

Consider this example:
```
#[tokio::main]
async fn main() {
    let (tx, mut rx1) = broadcast::channel(16);

    tokio::spawn(async move {
        let r = rx1.recv().await.unwrap();
        println!("Rx1 received: {}", r);
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
    });

    tx.send(20).unwrap();

    println!("Waiting for subscribers to drop");
    tx.closed().await;
    println!("Subscribers dropped");
}
```
Here the call to `closed` will block because it will only be polled once right when you call it. But the subscriber is not dropped at that point, so the `poll_fn` will return `Pending`. But because the future is never be polled again, this will not resolve after that.

Reply:
Indeed, this is not correct. When returning `Pending`, you must register for wakeups so that you are polled again when progress can be made.

Compare with the mpsc channel:

https://github.com/tokio-rs/tokio/blob/4825c444eb29b05c9c0b6f4a4c89475cb2556c65/tokio/src/sync/mpsc/chan.rs#L58-L59

Reply:
True, and I guess `poll_fn` is flagged as unstable in CI anyway. I'll find another way around this - thanks both.

Reply:
Tokio has a `poll_fn` implementation you can use if you find that you need it.